src/System.Diagnostics.Process/tests/ProcessTests.cs

```diff
@@ -1804,6 +1804,37 @@ public void TestLongProcessIsWorking()
             Assert.True(p.HasExited);
         }
 
+        [PlatformSpecific(~TestPlatforms.OSX)]
+        [Fact]
+        public void GetProcesses_LongProcessName()
+        {
+            string commandName = "sleep";
+
+            // sleep program doesn't exist on some flavor
+            if (!IsProgramInstalled(commandName))
+            {
+                return;
+            }
+
+            string longProcessName = "123456789012345678901234567890";
+            string sleepCommandPathFileName = Path.Combine(TestDirectory, longProcessName);
+            File.Copy(GetProgramPath(commandName), sleepCommandPathFileName);
+
+            // start sleep program and wait for some seconds
+            using (Process px = Process.Start(new ProcessStartInfo { FileName = sleepCommandPathFileName , Arguments = "30", UseShellExecute = true}))
+            {
+                try
+                {
+                    Assert.Contains(Process.GetProcesses(), p => p.ProcessName == longProcessName);
```

Comment:
`sleep` is available as a [`busybox` command](https://busybox.net/downloads/BusyBox.html) on Alpine. If we do `sleep 30 &\ ; ps uax`, we get `sleep 30` job running in background (or run ``jobs` after `sleep 30 &`). But if we follow the symlink (`/bin/sleep`), it points to `/bin/busybox`.

`readlink $(command -v sleep)` -> `/bin/busybox`

Test failure:
![image](https://user-images.githubusercontent.com/3840695/56420348-426c5580-62a6-11e9-957e-13dc53a1b8b1.png)

one of these `busybox` processes are `sleep`. We can assert with PID to confirm:

```c#
try
{
    Assert.Contains(Process.GetProcesses(), p => p.ProcessName == longProcessName);
}
catch(ContainsException)
{
    // on Alpine, sleep is a symlink which points to /bin/busybox
    Assert.Contains(Process.GetProcesses(), p => p.ProcessName == "busybox" && p.ProcessId == px.ProcessId);
}
```

or simply:


```c#
 // on Alpine, sleep is a symlink which points to /bin/busybox
Assert.Contains(Process.GetProcesses(), p => p.ProcessName == longProcessName || ( p.ProcessName == "busybox" && p.ProcessId == px.ProcessId));
```

Reply:
Maybe something more specific for Alpine run to avoid possible false positive. 
BTW I wonder why first time I run CI all works well(except OSX), maybe is a new CI leg.

Reply:
@am11 thank's!

Reply:
It still failed with [same error](https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F36913~2Fmerge/test~2Ffunctional~2Fcli~2F/20190419.5/workItem/System.Diagnostics.Process.Tests/analysis/xunit/System.Diagnostics.Tests.ProcessTests~2FGetProcesses_LongProcessName). Maybe we can try adding some temporary debug info in `catch` to find out some details:

```c#
var runninProcesses = Process.GetProcesses();
try
{
    Assert.Contains(runninProcesses, ...
}
catch(ContainsException)
{
    Assert.False(true, $"Expected ID {px.Id} in collection: [" + string.Join(" ,", runninProcesses.Select(p => $"{p.Name}({p.Id})")) + "]");
}

Reply:
tried to setup an Alpine image but I've got some issue with packages...try with CI

Reply:
I am also testing with Alpine (WSL):

```
apk update
apk add nano bash curl icu-libs git
curl -O https://dotnetcli.azureedge.net/dotnet/Sdk/master/dotnet-sdk-latest-linux-musl-x64.tar.gz
mkdir -p /usr/bin/dotnet
 tar -xzvf dotnet-sdk-latest-linux-musl-x64.tar.gz  -C /usr/bin/dotnet
```

Reply:
I get some error in `apk add git` I don't know why "Linux" is not my default working OS probably I'm missing something basic

Reply:
Locally I am able to reproduce it, not sure how the CI is passing with log commit now. 🤷‍♂️
When the process `/path/to/tempDirectory/123456789012345678901234567890` starts, we get error: `applet not found`. This error emerges from `busybox` which requires the applet name at  `argv[0]` to be a known command (`/path/to/tempDirectory/sleep` works).

Can we use something other than `sleep` (that is not a busybox command) for testing? Alternatively, we can skip this test for Alpine.

Reply:
> not sure how the CI is passing with log commit now. 🤷‍♂️

:confounded:

@stephentoub @danmosemsft @krwq should we disable for `Alpine`?

Reply:
Maybe it is worth trying:

<details>
  <summary>diff to be applied</summary>

```diff
             string longProcessName = "123456789012345678901234567890";
             string sleepCommandPathFileName = Path.Combine(TestDirectory, longProcessName);
-            File.Copy(GetProgramPath(commandName), sleepCommandPathFileName);
+
+            // in some flavors, sleep is a busybox command so we cannot copy it with different name straight up
+            if (!IsProgramInstalled("busybox"))
+            {
+                File.WriteLine(sleepCommandPathFileName, "#!/bin/sh\n\nexec busybox sleep $1");
+            }
+            else
+            {
+                File.Copy(GetProgramPath(commandName), sleepCommandPathFileName);
+            }

             // start sleep program and wait for some seconds
             using (Process px = Process.Start(new ProcessStartInfo { FileName = sleepCommandPathFileName , Arguments = "30", UseShellExecute = true}))
@@ -1826,12 +1835,7 @@ namespace System.Diagnostics.Tests
                 var runninProcesses = Process.GetProcesses();
                 try
                 {
-                    // on Alpine, sleep is a symlink which points to /bin/busybox
-                    Assert.Contains(runninProcesses, p => p.ProcessName == longProcessName || ( p.ProcessName == "busybox" && p.Id == px.Id));
-                }
-                catch(Xunit.Sdk.ContainsException)
-                {
-                    Assert.False(true, $"Expected ID {px.Id} in collection: [" + string.Join(" ,", runninProcesses.Select(p => $"{p.ProcessName}({p.Id})")) + "]");
+                    Assert.Contains(runninProcesses, p => p.ProcessName == longProcessName);
                 }
                 finally
                 {
```
</details>

i guess if a future flavor has tobox (like Android) instead of busybox, this test may need to be modified further.

Reply:
> Locally I am able to reproduce it, not sure how the CI is passing with log commit now. 🤷‍♂️
When the process /path/to/tempDirectory/123456789012345678901234567890 starts, we get error: applet not found. This error emerges from busybox which requires the applet name at argv[0] to be a known command (/path/to/tempDirectory/sleep works).

This would be because when it does `GetProcesses` it is lucky and gets the process before it exits with the error, so it is random. 

In your suggested approach did you mean `if (IsProgramInstalled("busybox"))` instead of `if (!IsProgramInstalled("busybox"))` ? But, can you not invoke `busybox` as a process - why is `sh` needed?

Reply:
Right, it should be without `!`. I think to skip this test on Alpine (or more specifically for busybox), we can check for: `IsProgramInstalled("busybox")` and exists. IMO, it would be good if this test executes a more sustainable process for all platforms? Maybe something that we can be implemented and added to https://github.com/dotnet/corefx-testdata.

> can you not invoke `busybox` as a process - why is `sh` needed?

```sh
cp /bin/sleep /home/sleep
/home/sleep # this is fine
/bin/busybox sleep # this is fine
/bin/busybox /bin/sleep # this is fine
/bin/busybox /home/sleep # this is fine

cp /bin/sleep /home/mysleep
/home/mysleep # this throws error: applet not found
/bin/busybox mysleep # also throws error: applet not found
/bin/busybox /home/mysleep # also throws error: applet not found
```

With shell script:

```
printf %s\\n '#!/bin/sh' 'exec busybox sleep $1' > /home/myscriptysleep && chmod +x /home/myscriptysleep
```

we can execute `/home/myscriptysleep 30 &`, but `ps uax` list it as:

```
 1788 root      0:00 busybox sleep 30
```

which defeats the purpose of this test, as you noted. 😅

Reply:
Did either of you verify manually that the fix works on Alpine?

Reply:
I have tested on Alpine and updated the diff patch (with a cleanup line in `finally`): https://github.com/dotnet/corefx/pull/36913#discussion_r277015612. Although invoked explicitly with `sh`, `GetProcesses` API (as does `ps uax` command) returns `ProcessName == busybox`, so i have changed `p.ProcessName.Contains` to `p.ProcessName ==`.

Reply:
@am11 not sure I follow, do you see a way to make a useful test (ie a test that verifies this fix) on Alpine? It does not seem that looking for busybox in the process name achieves this. Does it help if you copy busybox itself to a longer name?

If there is no easy way to test on Alpine, I'm fine checking this in without a test on Alpine, for now.

Reply:
> Does it help if you copy busybox itself to a longer name?

I think not; `cp /bin/busybox /tmp/123456789012345678901234567890` and executing with the long name results in the same `applet not found` error.

Is there anything other than `sleep` that could be used? We can compile a C program (like in `NativeLibrary` tests are doing), or use one of the existing executable native binary in test data that is compiled for the target OS if possible.

Reply:
I think it would be fine to open a separate issue to add a test for Alpine. Maybe using a custom library like NativeLibrary tests.

Reply:
@MarcoRossignoli, based on @danmosemsft's suggestion, we can revert changes made in 0c53c36, c8e1287 and cb11900 (or simply `git reset --hard 790de76`) and apply something like the following:

```diff
             // sleep program doesn't exist on some flavor
+            // in flavors such as Alpine, sleep is a busybox command which cannot be renamed
+            // TODO (##github-issue##): build a custom long-running executable for this test instead of sleep
-             if (!IsProgramInstalled(commandName))
+             if (!IsProgramInstalled(commandName) || IsProgramInstalled("busybox"))
```

Reply:
ping should run forever, right?

Reply:
perhaps `ping 127.0.0.1` can act as your long running command?

Reply:
@krwq good thinking, but ping is busybox too, unfortunately:
```
DESKTOP-CJJBOPC:~# readlink $(command -v sleep)
/bin/busybox
DESKTOP-CJJBOPC:~# readlink $(command -v ping)
/bin/busybox
```

Reply:
Ok will do with revert and skip on `busybox` system